### PR TITLE
(cloudwatch) alerting

### PR DESCRIPTION
**Pull request by mtanda**
Add alert feature for CloudWatch

### CODE REVIEW

**Comment by torkelo** [comment]
to merge this we need to move all the api/cloudwatch code into the tsdb package. So the same code is used for both alerting and for the frontend data source query. 

This means cloudwatch should not use the dataproxy anymore but the new `/api/tsdb/query` endpoint that is used by the the work in progress SQL data source (and test data source). 
https://github.com/grafana/grafana/blob/master/pkg/api/metrics.go#L19

This might require moving code from the frontend to the backend as well.

**Comment by torkelo** [comment]
Let me know if you have questions how this should work

**Comment by mtanda** [comment]
haha, I thought you would say that 😎 
I'll try to understand your work, and revise this PR!!

**Comment by mtanda** [comment]
@torkelo I want to know how I implement `metricFindQuery` in `tsdb` package.
Most of function in https://github.com/grafana/grafana/blob/master/pkg/api/cloudwatch/metrics.go, is `metricFindQuery`.

But `tsdb` query only support time series data. It is impossible to return metricFindQuery results.
https://github.com/grafana/grafana/blob/3ca3c9622696d5cce1b7d5c23f2056e97baae8fa/pkg/api/metrics.go#L48
https://github.com/grafana/grafana/blob/3ca3c9622696d5cce1b7d5c23f2056e97baae8fa/pkg/tsdb/request.go#L9
https://github.com/grafana/grafana/blob/3ca3c9622696d5cce1b7d5c23f2056e97baae8fa/pkg/tsdb/models.go#L27-L30
https://github.com/grafana/grafana/blob/3ca3c9622696d5cce1b7d5c23f2056e97baae8fa/pkg/tsdb/models.go#L47-L52

**Comment by torkelo** [comment]
That is true, we need a new api endpoint and new tsdb request for "MetricSeach" queries. 
Maybe POST `/api/tsdb/search/`

**Comment by mtanda** [comment]
@torkelo Do you implement it in v4.3?

**Comment by torkelo** [comment]
maybe, hope to continue work on MySQL and Postgres datasource next week and this feature is going to be needed by those as well (as they are built from the ground with all tsdb code in the backend)

**Comment by mtanda** [comment]
I wait for your work, and when the `/api/tsdb/search/` is available, continue moving code from frontend to backend.
The alert feature already works, if possible I want to merge this PR, and create another PR for the remain works.

@torkelo How do you think? Is it possible to merge this?

**Comment by mtanda** [comment]
(I want to get feedback from nightly build user.)

**Comment by torkelo** [comment]
I can understand that. Just afraid to merge this and be left with so much duplicated code that is never going to be consolidated.

**Comment by mtanda** [comment]
Of course, I promise removing duplicated code.

**Comment by timw077** [comment]
@torkelo @mtanda  Can we get an update?

**Comment by hflamboauto1** [comment]
@mtanda  Thanks for the work.
Now some months have gone away.. how usable is this still ? What do you think about a merge of last version into feature branch ?

**Comment by mtanda** [comment]
@hflamboauto1 I rebase and test it. It works.
This is usable. Please try this, and give feedback.

**Comment by boostrack** [comment]
Hi @mtanda 

Just have done it and  is running. 
Overall seems to work. Thanks for the work.

I am having trouble having the rules that should fire, not firing according to test output.

if i click the test button it shows that is supposed to fire .. all good.. but the triggering seems to not be happening to my slack channel.

File: public/app/plugins/datasource/cloudwatch/datasource.js
Comment (hflamboauto1) [inline public/app/plugins/datasource/cloudwatch/datasource.js]: you rock @mtanda

**Comment by deviousgeek** [comment]
Been using this a little bit, seems to be working well! Gon' merge this?

File: public/app/plugins/datasource/cloudwatch/datasource.js
Comment (golimer) [inline public/app/plugins/datasource/cloudwatch/datasource.js]: PLEASE merge asap

**Comment by golimer** [comment]
Guys please merge as soon as possible this PR

**Comment by Nokius** [comment]
@mtanda works well for me many thanks

**Comment by mtanda** [comment]
resolve conflicts.

**Comment by oleg-z** [comment]
@mtanda Have you tried to run it when CW datastore configured with Access Key and Secret Key? When I tried to run I had issues with corrupted access key and secret key. Encrypted datastore returned correct access key and secret key only for first item in batch. All subsequent requests returned complete garbage keys or partially correct keys. 

I isolated issue to concurrent execution in tsdb/cloudwatch/cloudwatch.go#Execute. I solved it by moving getting access/secret key from GetClient() directly to Execute and saving them as global vars. I was just trying to make it work so it's not ideal solution but I'm curios if you seen/tested such behavior.

**Comment by mtanda** [comment]
@oleg-z Sorry I didn't test it.
I'm not sure why the concurrent execution cause such issue.
Past query implementation doing same thing, but no one report the issue.
(I add new commit, caching creds, but it may not fix the issue.)

**Comment by duke-cliff** [comment]
@mtanda Thanks man. We almost gave up Grafana until we saw this PR. Cannot imagine why so far Grafana does not support cloudwatch alerting.

**Comment by duke-cliff** [comment]
@mtanda  It works fine with the 5.0.0 preversion's own graph plugin. But the thing is we have an existing setup of 30+ dashboards with multiple panels. I tried to reuse the existing grafana.db file, it will raise metrics request errors on all cloudwatch datasources. If I rebuild a graph using the exact same query, the new panel will be fine. Do you know how to fix these upgrade issues.

**Comment by golimer** [comment]
possible to merga to 4.5 version? :/

**Comment by duke-cliff** [comment]
@mtanda Another bug report: I cannot query the graph with 24 hours range, it will return this error:InvalidParameterCombination: You have requested up to 1,455 datapoints, which exceeds the limit of 1,440. You may reduce the datapoints requested by increasing Period, or decreasing the time range.   The 4.x works fine with all kinds of date range.

**Comment by mtanda** [comment]
@duke-cliff Thanks for feedback!

I try to replicate the broken dashboard issue.
I rebased to current master, and build grafana, create dasshboard in current master.
https://github.com/grafana/grafana/commit/40c008f87075db21fe79cecad86bbc28b7157f69
And then, switch to cloudwatch_alert branch, and build grafana, and open the dashboard.

It works. I'm not sure why your dashboard is broken, if other change is affected, you can test it by using grafana nightly build.

**Comment by mtanda** [comment]
About error when 24 hour time range, it might be caused by this line.

https://github.com/grafana/grafana/pull/8050/files#diff-ed5cc68bef51197743c5863b078db25bR159
>  StartTime:  aws.Time(startTime.Add(-time.Minute * 15)), 

CloudWatch metrics output is sometimes delayed, most of the delay is 1-3 min, but in the worst case, the delay could be 15 min.
To ensure, Grafana can get some metrics value, I add 15min lookback window.

It is difficult problem... try to solve it.

**Comment by mtanda** [comment]
image
By specifying query range more than 15min, Grafana can handle metrics delay.
Simply remove the lookback window.

**Comment by duke-cliff** [comment]
@mtanda The broken dashboards were complaining bad tokens in the log. I will try to reproduce it later and give you more information.

**Comment by duke-cliff** [comment]
@mtanda EROR[09-05|17:37:47] Metric request error                     logger=context userId=0 orgId=1 uname= error="InvalidClientTokenId: The security token included in the request is invalid.\n\tstatus code: 403, request id: ee6aecca-9260-11e7-ac1f-e509368efd40"

**Comment by duke-cliff** [comment]
@mtanda EROR[09-05|17:39:39] Metric request error                     logger=context userId=1 orgId=1 uname=admin error="time: unknown unit d in duration -1d"

**Comment by duke-cliff** [comment]
@mtanda EROR[09-05|17:45:01] Alert Rule Result Error                  logger=alerting.evalHandler ruleId=1 name="Response Time alert" error="tsdb.HandleRequest() error strconv.Atoi: parsing \"1s\": invalid syntax" changing state to=alerting

**Comment by duke-cliff** [comment]
@mtanda The data point over flow issue is fixed.

**Comment by mtanda** [comment]
> EROR[09-05|17:45:01] Alert Rule Result Error logger=alerting.evalHandler ruleId=1 name="Response Time alert" error="tsdb.HandleRequest() error strconv.Atoi: parsing "1s": invalid syntax" changing state to=alerting

I fix the bug for this error.

**Comment by mtanda** [comment]
> EROR[09-05|17:37:47] Metric request error logger=context userId=0 orgId=1 uname= error="InvalidClientTokenId: The security token included in the request is invalid.\n\tstatus code: 403, request id: ee6aecca-9260-11e7-ac1f-e509368efd40"

This is replicated in my env, when I use access key. Try to fix it.

**Comment by mtanda** [comment]
https://github.com/grafana/grafana/pull/8050#issuecomment-327709297

Sorry, I was misunderstanding.
I forgot to remove another test alert which configuration is not valid.
When I add new CloudWatch datasource with access key, and add alert, it works.

**Comment by mtanda** [comment]
@duke-cliff Would you re-enter your access key / secret key to the datasource setting, and confirm bug is still occurred?

**Comment by bergquist** [comment]
Great work.

But the only way will merge this is if we remove the cloudwatch code from the dataproxy. It might seems harsh but we really want to avoid the code duplication. In the other datasource we dont have that option but here we do. 

regarding an api like /api/tsdb/search/ , we do not think this is required as the existing query http api can return table formatted data that can be used for metricFind style queries (see MySQL data source for reference)

**Comment by mtanda** [comment]
make progress.
I hope nobody change current master's cloudwatch code... (because hard to rebase)

**Comment by mtanda** [comment]
@bergquist I implement metricFindQuery() which use table format.
Please check implementation if you have spare time.

**Comment by mtanda** [comment]
I need to move `pkg/api/cloudwatch/cloudwatch.go` to `pkg/tsdb/cloudwatch`.
The code implement annotation query, I find the way to change response format to table data.

**Comment by mtanda** [comment]
@bergquist 
>  if we remove the cloudwatch code from the dataproxy.

If I remove the dataproxy code in this week, when is the code merge to the master?
I don't know Grafana branch policy, if another 4.x release will coming, is it hard to merge to master until all 4.x release is finished?
I guess, you already merge your breaking change (https://github.com/grafana/grafana/pull/9187) to master, it is not impossile...

**Comment by torkelo** [comment]
it can be merged if it's stable, and well tested :)

**Comment by mtanda** [comment]
@torkelo Okay, I'll finish the rest of work until this weekend.

**Comment by mtanda** [comment]
I need to wait the tsdb refactoring...
It's very hard way until merge.
https://github.com/grafana/grafana/commits/tsdb_rewrite

**Comment by bergquist** [comment]
@mtanda sorry! I forgot your working on this :( the api should be stable soon. The reason we are working on this is to make it easier to add support for backend plugins https://github.com/grafana/grafana/issues/6841
I'll notify you when we feel happy about it. 

once again sorry. If there are any new changes they will be minor.

**Comment by audip** [comment]
Eagerly waiting for this PR to get merged in. Great work @mtanda

**Comment by bergquist** [comment]
@mtanda after https://github.com/grafana/grafana/commit/ed661767f846601124ba0ec349f0ad10216764ef we wont make any changes to the signature. Eventually we will replace the request/response structs with protobuf structs. But that work will happen in a branch initially.

**Comment by mtanda** [comment]
@bergquist Sorry for complaining.
You have a task, which should finish ASAP, I can't stop it.

But, keep rebasing and resolving conflict is a very hard thing for me.
I consider to give up, but I already use a lot of time, so I want to avoid the time becomes sunk costs.
In honest, this PR is not for me, there are many requirements for alerting, but nobody will implement it, so this is voluntary work...

If the PR is merged to master before your refactoring start and brush it up in the master branch, my work will progress more easily.
I propose the merge process before. See the comment https://github.com/grafana/grafana/pull/8050#issuecomment-293496628.
I promise I remove duplicative code, but the request is refused...

**Comment by bergquist** [comment]
Ill look into this PR tomorrow morning.

File: pkg/tsdb/cloudwatch/cloudwatch.go
Comment (bergquist) [inline pkg/tsdb/cloudwatch/cloudwatch.go]: This might cause problems since model is a pointer to queries[i]. when the loop iterates model will be updated. 
In this case Get() will return a new value so it will work as expected but I think its better to be explicit. like the refId sent to this function.

File: pkg/tsdb/cloudwatch/types.go
Comment (bergquist) [inline pkg/tsdb/cloudwatch/types.go]: I think it would be better to include the type in this struct and parse it all at once.

**Review by bergquist — COMMENTED** [review]
You come a long way of supporting cloudwatch alerting. But I still experience some problems. 

I solved the conflicts and pushed it to a new branch `cloudwatch_alerting`. I don't know how to commit to your PR. or If its even possible.

the commit https://github.com/grafana/grafana/commit/9c693af9a6944d1164147e74532a693bb085505c will solve the conflicts. 

You can either cherry-pick that or keep working against `cloudwatch_alerting` its up to you :)

If you want to keep working on this PR I will review it regularly.

**Comment by mtanda** [comment]
@bergquist Thank you for taking time for me.
I don't want to bother you.
But I sometimes I think, is there more effective contribution process.
Grafana core is changed drastically, it is hard to keep long living PR.
On the other hand, keep Grafana stable is hard if PR is merged quickly.
We need to consider how we can improve the process.

I cherry-pick your refactoring commit.
I need to resolve one conflict, so I keep to work on this PR.

I change query parameter, it may break alert feature.
I concentrate to porting all metric find query, I didn't notice...
I need to use same query parameter format in frontend code, and alerting.

Thanks again for your check, I'll fix it.

**Comment by mtanda** [comment]
My porting work is mostly done.
Check the behavior, and it seems work.

**Comment by torkelo** [comment]
that's great @mtanda we will try to find time to review & merge this soon! thanks for all your hard work, I know many are going to love this.

**Comment by mtanda** [comment]
@torkelo Thanks. I'm waiting for feedback!

File: pkg/tsdb/cloudwatch/cloudwatch.go
Comment (mtanda) [reply pkg/tsdb/cloudwatch/cloudwatch.go]: I forgot to reply here.
This is fixed.

File: pkg/tsdb/cloudwatch/cloudwatch.go
Comment (mtanda) [reply pkg/tsdb/cloudwatch/cloudwatch.go]: This is fixed by you. Thanks.

File: pkg/tsdb/cloudwatch/types.go
Comment (mtanda) [reply pkg/tsdb/cloudwatch/types.go]: I don't include the type, because there are many query type, and I don't want to prepare `struct` for the each query.

File: public/app/plugins/datasource/cloudwatch/datasource.js
Comment (bergquist) [inline public/app/plugins/datasource/cloudwatch/datasource.js]: CloudWatchDatasource also needs to implement `targetContainsTemplate`. Its used to display an error if query used by alerting contains a template variable. Ex https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/graphite/datasource.ts#L140

**Review by bergquist — CHANGES_REQUESTED** [review]

File: public/app/plugins/datasource/cloudwatch/datasource.js
Comment (mtanda) [reply public/app/plugins/datasource/cloudwatch/datasource.js]: Sorry, I didn't notice. Add it!

**Review by bergquist — APPROVED** [review]

**Comment by bergquist** [comment]
@mtanda Ive tested this and it looks good. Great work! Anything else you would like to do in this PR before I merge it? :)

**Comment by mtanda** [comment]
@bergquist Nothing, it is ready to merge!

**Comment by mtanda** [comment]
Many thanks!!

**Comment by golimer** [comment]
so if merged maybe will be released with new release?

**Comment by efrat-arad** [comment]
I've installed Grafana 4.6.2 and defined alert for CloudWatch metric.
I'm receiving the below error all the time when an alert exist (if the alert is removed I do not see this error):
Error: tsdb.HandleRequest() error Could not find executor for data source type: cloudwatch
Alerting
Any idea why this error appears?

**Comment by mtanda** [comment]
I don't have an idea... I can't replicate it.

**Pull request merged**
